File: src/ros2_bridge/mappers/diagnostic_msgs/diagnostic_array.rs
//! Typed mapper for `diagnostic_msgs/msg/DiagnosticArray`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn diagnostic_array_to_bus(msg: ros_env::diagnostic_msgs::msg::DiagnosticArray) -> crate::diagnostic_msgs::msg::v1::DiagnosticArray {
    crate::diagnostic_msgs::msg::v1::DiagnosticArray {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        status: msg.status.into_iter().map(crate::ros2_bridge::mappers::diagnostic_msgs::diagnostic_status::diagnostic_status_to_bus).collect(),
    }
}

pub(crate) fn diagnostic_array_to_ros(bus: crate::diagnostic_msgs::msg::v1::DiagnosticArray) -> ros_env::diagnostic_msgs::msg::DiagnosticArray {
    ros_env::diagnostic_msgs::msg::DiagnosticArray {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        status: bus.status.into_iter().map(crate::ros2_bridge::mappers::diagnostic_msgs::diagnostic_status::diagnostic_status_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct DiagnosticMsgsDiagnosticArrayMapper;

impl TypedTopicMapper for DiagnosticMsgsDiagnosticArrayMapper {
    type Ros = ros_env::diagnostic_msgs::msg::DiagnosticArray;
    type Bus = crate::diagnostic_msgs::msg::v1::DiagnosticArray;

    fn type_name(&self) -> &'static str {
        "diagnostic_msgs/msg/DiagnosticArray"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(diagnostic_array_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(diagnostic_array_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/diagnostic_msgs/diagnostic_array.py
"""Generated mapper for `diagnostic_msgs/msg/DiagnosticArray`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.diagnostic_msgs.diagnostic_status import diagnostic_status_to_bus, diagnostic_status_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.diagnostic_msgs.msg.v1 import DiagnosticArray as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def diagnostic_array_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.status.extend([diagnostic_status_to_bus(x) for x in msg.status])
    return bus


def diagnostic_array_to_ros(bus):
    from diagnostic_msgs.msg import DiagnosticArray as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.status = [diagnostic_status_to_ros(x) for x in bus.status]
    return out


class DiagnosticMsgsDiagnosticArrayMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from diagnostic_msgs.msg import DiagnosticArray as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return diagnostic_array_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return diagnostic_array_to_ros(bus)
